src/USER-HMA/compute_hma.cpp

```diff
@@ -0,0 +1,473 @@
+
+/* ---------------------------------------------------------------------- */
+
+ComputeHMA::ComputeHMA(LAMMPS *lmp, int narg, char **arg) :
+  Compute(lmp, narg, arg), id_temp(NULL), deltaR(NULL)
+{
+  if (narg < 4) error->all(FLERR,"Illegal compute hma command");                    
+  if (igroup) error->all(FLERR,"Compute hma must use group all");
+  if (strcmp(arg[3],"NULL") == 0) {error->all(FLERR,"fix ID specifying the set temperature of canonical simulation is required");}       
+  else {
+    int n = strlen(arg[3]) + 1;                  
+    id_temp = new char[n];   
+    strcpy(id_temp,arg[3]);                         
+  }
+ 
+  create_attribute = 1;                  
+  extscalar = 1;                         
+  timeflag = 1;                         
+
+  // (from compute displace/atom) create a new fix STORE style  
+  // our new fix's id (id_fix)= compute-ID + COMPUTE_STORE 
+  // our new fix's group = same as compute group
+
+  int n = strlen(id) + strlen("_COMPUTE_STORE") + 1;
+  id_fix = new char[n];
+  strcpy(id_fix,id);
+  strcat(id_fix,"_COMPUTE_STORE");
+
+  char **newarg = new char*[6];     
+  newarg[0] = id_fix;
+  newarg[1] = group->names[igroup];
+  newarg[2] = (char *) "STORE";              
+  newarg[3] = (char *) "peratom";
+  newarg[4] = (char *) "1";
+  newarg[5] = (char *) "3";
+  modify->add_fix(6,newarg);        
+  fix = (FixStore *) modify->fix[modify->nfix-1];    
+ 
+  delete [] newarg;                               
+
+  // calculate xu,yu,zu for fix store array
+  // skip if reset from restart file
+
+  if (fix->restart_reset) fix->restart_reset = 0;   
+  else {
+    double **xoriginal = fix->astore;            
+    double **x = atom->x;
+    imageint *image = atom->image;
+    int nlocal = atom->nlocal;
+
+    for (int i = 0; i < nlocal; i++)
+      domain->unmap(x[i],image[i],xoriginal[i]);          
+  }
+
+  vector_flag = 1;
+  extvector = -1;
+  comm_forward = 0; // 3 if 2nd derivative needed
+
+  computeU = computeP = computeCv = -1;
+  returnAnharmonic = 0;
+  std::vector<int> extvec;
+  for (int iarg=4; iarg<narg; iarg++) {
+    if (!strcasecmp(arg[iarg], "u")) {
+      computeU = extvec.size();
+      extvec.push_back(1);
+    }
+    else if (!strcasecmp(arg[iarg], "p")) {
+      if (iarg+2 > narg) error->all(FLERR,"Illegal fix hma command");
+      computeP = extvec.size();
+      deltaPcap = force->numeric(FLERR, arg[iarg+1]);
+      extvec.push_back(0);
+      iarg++;
+    }
+    else if (!strcasecmp(arg[iarg], "cv")) {
+      computeCv = extvec.size();
+      comm_forward = 3;
+      extvec.push_back(1);
+    }
+    else if (!strcasecmp(arg[iarg], "anharmonic")) {
+      // the first time we're called, we'll grab lattice pressure and energy
+      returnAnharmonic = -1;
+    }
+    else {
+      error->all(FLERR,"Illegal fix hma command");
+    }
+  }
+
+  if (extvec.size() == 0) {
+    error->all(FLERR,"Illegal fix hma command");
+  }
+  size_vector = extvec.size();
```

Comment:
why so complicated? why not collect the data in `extvec` away, it could be allocated to a reasonable default size with `memory->smalloc()` and then grown in chunks as needed with `memory->grow()`.

Reply:
Right, I can see that this could be simplified a bit, but I'm not sure what you're specifically suggesting.  I could allocate `extlist` to be of size 3 (maximum size it should ever be).  I'd still need to keep track of how many items it actually contained (to set `computeU`, `computeP`, `computeCv`), but I could drop `extvec` entirely.

Reply:
you can just define `int size_vector=0` and use it to index `extlist[]` and increment afterwards. so at the end it will be the size. the same is done with lists of items in LAMMPS all over the place.